corehq/apps/couch_sql_migration/tests/test_migration.py

```diff
@@ -1397,6 +1432,44 @@ def test_migrate_ledgers(self):
 
         self._compare_diffs([])
 
+    def test_migrate_partially_migrated_form_with_ledger(self):
+        case_id = "test-case"
+        self.submit_form(TEST_FORM, timedelta(days=-5))  # create test-case
+        form1 = self._set_balance(case_id, {
+            self.liquorice: 50,
+            self.sherbert: 100,
+        }, timedelta(days=-3))
+        form2 = self._set_balance(case_id, {self.liquorice: 75}, timedelta(days=-1))
+        print("ledger forms:", form1, form2)
+        with self.skip_case_and_ledger_updates(form1):
```

Comment:
what happens if you skip `form2`, does the stock end up in the correct state with a balance of 75? I think it will not based on the current code in `_iter_missing_ledgers`

Reply:
I tried this locally and was surprised to find that it didn't give any stock diffs. Digging into it more I found that it was hitting this: https://github.com/dimagi/commcare-hq/blob/082acc9bd7121a1047ef35a4c16aea0fdbd7698a/corehq/apps/couch_sql_migration/casediff.py#L193-L196

It seemed odd that there would be duplicate transaction but digging deeper I found a bug in `has_duplicate_transactions`. This method compares the number of DB transactions with the number in the form. (get read to cringe?) This doesn't work because the legacy backend creates an additional transaction in certain cenarios:

https://github.com/dimagi/commcare-hq/blob/579b52f1037d30c8e9184214f7278bb1d6a8b9d2/corehq/ex-submodules/casexml/apps/stock/signals.py#L28-L47

With this change to `StockTransactionLoader.iter_stock_transactions` I got the diffs I was expecting:

```diff
     def iter_stock_transactions(self, form_id):
         xform = FormAccessorCouch.get_form(form_id)
         assert xform.domain == _diff_state.domain, xform
         for report in get_all_stock_report_helpers_from_form(xform):
             for tx in report.transactions:
                 yield report.report_type, tx
+                if tx.action == TRANSACTION_TYPE_STOCKONHAND:
+                    yield report.report_type, tx
```

(I think this change is all that is necessary since they are only used to count the number of transactions).

Reply:
@snopoke I think the concerns you raise are resolved by https://github.com/dimagi/commcare-hq/pull/26768/commits/fc221210fd95a8b05a5d8a03cbe3e6310d3093cf Please verify that the test I added matches what you meant by "what happens if you skip `form2`?"

Reply:
yes, that resolves it. I think the rebuild was the missing piece that was causing diffs even after fixing the comparison.